pkgs/dart_mcp/lib/src/api/tools.dart

```diff
@@ -197,6 +197,78 @@ enum JsonType {
   final String typeName;
 }
 
+/// Enum representing the types of validation failures when checking data
+/// against a schema.
+enum ValidationErrorType {
+  // General
+  typeMismatch,
+
+  // Schema combinators
+  allOfNotMet,
+  anyOfNotMet,
+  oneOfNotMet,
+  notConditionViolated,
+
+  // Object specific
+  requiredPropertyMissing,
+  additionalPropertyNotAllowed,
+  minPropertiesNotMet,
+  maxPropertiesExceeded,
+  propertyNamesInvalid,
+  propertyValueInvalid,
+  patternPropertyValueInvalid,
+  unevaluatedPropertyNotAllowed,
+
+  // Array/List specific
+  minItemsNotMet,
+  maxItemsExceeded,
+  uniqueItemsViolated,
+  itemInvalid,
+  prefixItemInvalid,
+  unevaluatedItemNotAllowed,
+
+  // String specific
+  minLengthNotMet,
+  maxLengthExceeded,
+  patternMismatch,
+
+  // Number/Integer specific
+  minimumNotMet,
+  maximumExceeded,
+  exclusiveMinimumNotMet,
+  exclusiveMaximumExceeded,
+  multipleOfInvalid,
+}
+
+/// A validation error with detailed information about the location of the
+/// error.
+extension type ValidationError.fromMap(Map<String, Object?> _value) {
+  factory ValidationError(
+    ValidationErrorType error, {
+    List<String>? path,
+    String? details,
+  }) => ValidationError.fromMap({
+    'error': error,
+    if (path != null) 'path': path,
+    if (details != null) 'details': details,
+  });
+
+  /// The type of validation error that occurred.
+  ValidationErrorType? get error => _value['error'] as ValidationErrorType?;
+
+  /// The path to the object that had the error.
+  List<String>? get path => _value['path'] as List<String>?;
```

Comment:
When reading from the map, if this is in a deserialized form, it will be a `List<Object?>`, so we need to cast it.

```suggestion
  List<String>? get path => (_value['path'] as List?)?.cast<String>();
```